### PR TITLE
[Merged by Bors] - feat: port RingTheory.Ideal.Quotient

### PR DESCRIPTION
**Pull request by jcommelin**
---
The problem here is that Lean cannot synthesize the instance `Module R R`, even though there is an instance for this in `Algebra.Module.Basic` called `Semiring.toModule` (leanprover/lean4#2074)

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by ChrisHughes24** [comment]
The problem here is that Lean cannot synthesize the instance `Module R R`, even though there is an instance for this in Algebra.Module.Basic called `Semiring.toModule`

**Comment by LukasMias** [comment]
This builds now, but needs the `set_option synthInstance.etaExperiment true` line, which we have agreed not to merge into mathlib currently. Otherwise `Module R R` is not found, as described above.

File: Mathlib/Data/UInt.lean [hunk 1]
```diff
@@ -88,15 +89,15 @@ run_cmd
...
       instance : Ring $typeName where
-        sub_eq_add_neg _ _ := congrArg mk (sub_eq_add_neg _ _)
-        add_left_neg _ := congrArg mk (add_left_neg _)
+        toAddCommGroup := inferInstanceAs (AddCommGroup $typeName)
         intCast n := ⟨n⟩
         intCast_ofNat _ := rfl
         intCast_negSucc _ := rfl
+        __ := inferInstanceAs (Semiring $typeName)
 
       instance : CommRing $typeName where
```
Comment (eric-wieser) [inline Mathlib/Data/UInt.lean:102]: Do you expect #2567 to make this better or worse?

**Comment by jcommelin** [comment]
The solution is to use `etaExperiment` in mathlib, but sparingly.

File: Mathlib/RingTheory/Ideal/Quotient.lean [hunk 2]
```diff
@@ -111,14 +113,17 @@ protected theorem eq : mk I x = mk I y ↔ x - y ∈ I :=
...
-  rfl
+theorem mk_eq_mk (x : R) : (Submodule.Quotient.mk x : R ⧸ I) = mk I x := rfl
 #align ideal.quotient.mk_eq_mk Ideal.Quotient.mk_eq_mk
 
 theorem eq_zero_iff_mem {I : Ideal R} : mk I a = 0 ↔ a ∈ I :=
   Submodule.Quotient.mk_eq_zero _
 #align ideal.quotient.eq_zero_iff_mem Ideal.Quotient.eq_zero_iff_mem
 
+-- Porting note: new theorem
+theorem mk_eq_mk_iff_sub_mem (x y : R) : mk I x = mk I y ↔ x - y ∈ I := by
```
Comment (alexjbest) [inline Mathlib/RingTheory/Ideal/Quotient.lean:124]: @jcommelin do you remember why you added this, it looks the same as Ideal.Quotient.eq with different implicitness of arguments?

**Pull request closed without merging**